File: shamelagpt-ios/shamelagptUITests/Helpers/TestDiagnostics.swift
import Foundation
import XCTest

struct UITestDiagnosticEvent {
    let testName: String
    let platform: String
    let locale: String
    let selectorOrTag: String
    let scenarioID: String
    let observedState: String
    let failureClass: String

    func asDictionary() -> [String: String] {
        [
            "test_name": testName,
            "platform": platform,
            "locale": locale,
            "selector_or_tag": selectorOrTag,
            "scenario_id": scenarioID,
            "observed_state": observedState,
            "failure_class": failureClass
        ]
    }
}

enum UITestDiagnostics {
    static func emit(_ event: UITestDiagnosticEvent) {
        guard let data = try? JSONSerialization.data(withJSONObject: event.asDictionary(), options: [.sortedKeys]),
              let json = String(data: data, encoding: .utf8)
        else {
            print("UITEST_DIAGNOSTIC: failed_to_encode")
            return
        }
        print("UITEST_DIAGNOSTIC: \(json)")
    }

    static func currentPlatformLabel() -> String {
        let env = ProcessInfo.processInfo.environment
        let model = env["SIMULATOR_MODEL_IDENTIFIER"] ?? "unknown_model"
        let runtime = env["SIMULATOR_RUNTIME_VERSION"] ?? "unknown_runtime"
        return "ios_simulator:\(model)|\(runtime)"
    }
}

File: shamelagpt-android/app/src/test/java/com/shamelagpt/android/testing/TestDiagnostics.kt
package com.shamelagpt.android.testing

/**
 * Standardized diagnostics payload for test failures.
 */
data class TestDiagnosticEvent(
    val testName: String,
    val platform: String,
    val locale: String,
    val selectorOrTag: String,
    val scenarioId: String,
    val observedState: String,
    val failureClass: String
)

object TestDiagnostics {
    fun emit(event: TestDiagnosticEvent) {
        val json = """{"test_name":"${event.testName}","platform":"${event.platform}","locale":"${event.locale}","selector_or_tag":"${event.selectorOrTag}","scenario_id":"${event.scenarioId}","observed_state":"${event.observedState}","failure_class":"${event.failureClass}"}"""
        println("TEST_DIAGNOSTIC: $json")
    }

    fun assertWithDiagnostics(
        condition: Boolean,
        event: TestDiagnosticEvent,
        assertionMessage: String
    ) {
        if (condition) {
            return
        }
        emit(event)
        throw AssertionError(assertionMessage)
    }
}
